src/Microsoft.Tye.Hosting/TokenReplacement.cs

```diff
@@ -58,6 +83,41 @@ private static HashSet<string> GetTokens(string text)
 
             return tokens;
         }
+        
+        private static (EffectiveBinding binding, string? text)? ResolveEnvironmentToken(string token, List<EffectiveBinding> bindings)
+        {
+            // The language we support for tokens is meant to be pretty DRY. It supports a few different formats:
+            //
+            // - ${host}: only allowed inside a connection string, it can refer to the binding.
+            // - ${env:SOME_VAR}: allowed anywhere. It can refer to any environment variable defined for *this* service.
+            // - ${service:myservice:port}: allowed anywhere. It can refer to the protocol/host/port of bindings.
+           
+            var keys = token[2..^1].Split(':');
+            if (keys.Length == 3 && keys[0] == "service")
+            {
+                var binding = bindings.FirstOrDefault(b => b.Service == keys[1] && b.Name == null)!;
+                return (binding, GetValueFromBinding(binding, keys[2]));
+            }
+            else if (keys.Length == 4 && keys[0] == "service")
+            {
+                var binding  = bindings.FirstOrDefault(b => b.Service == keys[1] && b.Name == keys[2])!;
+                return (binding, GetValueFromBinding(binding, keys[3]));
+            }
+
+            return null;
+
+            string? GetValueFromBinding(EffectiveBinding binding, string key)
+            {
+                return key switch
+                {
+                    "protocol" => binding.Protocol,
+                    "host" => binding.Host,
+                    "port" => binding.Port?.ToString(CultureInfo.InvariantCulture),
+                    "connectionString" => binding.ConnectionString,
+                    _ => null,
+                };
+            }
+        }
 
```

Comment:
This file isn't DRY or sexy... it's a bit of a copy and paste - but any suggestions I'm all ears!